earthfile2llb/converter.go

```diff
@@ -1300,8 +1300,16 @@ func (c *Converter) FinalizeStates(ctx context.Context) (*states.MultiTarget, er
 }
 
 // ExpandArgs expands args in the provided word.
-func (c *Converter) ExpandArgs(word string) string {
-	return c.varCollection.Expand(word)
+func (c *Converter) ExpandArgs(ctx context.Context, word string) (string, error) {
+	if !variables.ContainsShell(word) {
+		return c.varCollection.Expand(word), nil
+	}
+	pncvf := c.processNonConstantBuildArgFunc(ctx)
```

Comment:
This doesn't feel right - this will replace the **entire** value, not just the `$(...)` part.

The code that detects whether a word contains shell, needs to be expanded to also detect which part of that to replace.

Reply:
it should keep the complete string in tact, here's a test I added https://github.com/earthly/earthly/pull/1699/files#diff-b3414e5141e314866d254a961baf56400082c346f72c0633d94b3ec745dfd93bR65-R68

it does "hello$(cat /data)" which ends up "helloworld".

Reply:
Oh I see - it works because they are part of different words, I think. There may be edge cases where they are in the same word, and that will break it.